content/blog/2022-05-08-Apache-ShenYu(incubating)-Integrated-Skywalking-Practice-Observability/index.md

```diff
@@ -0,0 +1,130 @@
+- **LocalSpan transit.**
+  Other Rpc class plugins do not receive snapshots for concatenation like Asynchronous WebClient. Although you can modify other Rpc plugins to receive snapshots for concatenation, it is not recommended or necessary to do so.
+  This can be achieved by creating a LocalSpan in the thread where the ExitSpan is created, and then connecting the asynchronous EntrySpan and LocalSpan by `snapshot passing`. This can be done without changing the original plugin code.
+
+
+The span connection is shown below:  
+![](span-connect.png)
+
+You may ask if it is possible to create LocalSpan inside a generic plugin, instead of creating one separately for ShenYu Rpc plugin?
+The answer is no, because you need to ensure that LocalSpan and ExitSpan are in the same thread, and ShenYu is fully linked asynchronously. The code to create LocalSpan is reused in the implementation.
+
+
+## 3. Adding generalized call tracking to the gRPC plugin and keeping it compatible
+
+The existing SkyWalking gRPC plugin only supports calls initiated by way of stubs. For the gateway there is no proto file, the gateway takes generalized calls (not through stubs), so tracing rpc requests, you will find that the link will break at the gateway node.
+In this case, it is necessary to make the gRPC plugin support generalized calls, while at the same time needing to remain compatible and not affect the original tracing method. This is achieved by determining whether the request parameter is a DynamicMessage, and if it is not, then the original tracing logic through the stub is used.
+If not, then the original tracing logic via stubs is used, and if not, then the generalized call tracing logic is used. The other compatibility is the difference between the old and new versions of grpc, as well as the compatibility of various cases of obtaining server-side IP, for those interested in the source code.
+
+![](grpc-generic-call.png)
+
+## 4. ShenYu Gateway Observability Practice
+The above explains the principle of SkyWalking ShenYu plug-in implementation, the following deployment of applications to see the effect. SkyWalking powerful, in addition to the link tracking requires the development of plug-ins, other powerful features out of the box.
+Here only describe the link tracking and application performance analysis part, if you want to experience the power of SkyWalking features, please refer to the [SkyWalking official documentation](https://skywalking.apache.org/)
+
+### 4.1 Deploying Applications
+
+- After packaging SkyWalking Agent (` . /mvnw clean package -Pall -DskipTests`, recommended to download directly from [official website](https://skywalking.apache.org/downloads/) ), for the tracking of the shenyu gateway, put the shenyu plugin into the plugins directory
```

Comment:
I think this is unnecessary.
We directly refer to official docs of skywalking-agent and said how to activate the shenyu plugin.

Reply:
Make sense, also this reminds me, `building from source codes` is really rare for end users. Mostly they would choose an official release. You could mention 8.11.0, I think I may initial a release in May or June for it. Java agent is on the routine release stage.

Reply:
Very good advice, I have modified, thanks